src/server.c

```diff
@@ -249,6 +249,63 @@ bool user_requires_auth_query(PgUser *user)
 	return cf_auth_type != AUTH_TRUST && !user->from_auth_file;
 }
 
+void user_passwd_free(struct AANode *node, void *arg)
+{
+	PgUserPassword *user_passwd = container_of(node, PgUserPassword, tree_node);
+	free(user_passwd);
+	user_passwd = NULL;
+}
+
+/* compare string with PgUserPassword->username, for usage with btree */
+int user_passwd_node_cmp(uintptr_t user_passwd_ptr, struct AANode *node)
+{
+	const char *name = (const char *)user_passwd_ptr;
+	PgUserPassword *user_passwd = container_of(node, PgUserPassword, tree_node);
+	return strcmp(name, user_passwd->username);
+}
+
+char *user_password(const PgUser *user, const PgDatabase *db)
+{
+	PgUserPassword *user_passwd = NULL;
+	struct AANode *node;
+
+	if (db == NULL || user == db->forced_user || user->from_auth_file)
+		return (char *)user->passwd;
+
+	node = aatree_search((struct AATree *)&db->user_passwds, (uintptr_t)user->name);
```

Comment:
Cast away from const* here. Alternatively, can add `const*` modifier to signature `aatree_search`.